packages/shell-api/src/bulk.ts

```diff
@@ -0,0 +1,217 @@
+  _innerFind: any;
+  _parentBulk: Bulk;
+  _hint: any;
+  _arrayFilters: any;
+  constructor(innerFind: any, parentBulk: Bulk) {
+    this._innerFind = innerFind;
+    this._parentBulk = parentBulk;
+  }
+
+  _asPrintable(): string {
+    return 'BulkFindOp';
+  }
+
+  // Blocked by NODE-2757
+  collation(): void {
+    throw new MongoshUnimplementedError(
+      'collation method on fluent Bulk API is not currently supported. ' +
+      'As an alternative, consider using the \'db.collection.bulkWrite(...)\' helper ' +
+      'which accepts \'collation\' as a field in the operations.'
+    );
+  }
+
+  // Blocked by NODE-2751
+  arrayFilters(): BulkFindOp {
+    throw new MongoshUnimplementedError(
+      'arrayFilters method on fluent Bulk API is not currently supported.'
+    );
+  }
+
+  hint(hintDoc: Document): BulkFindOp {
+    assertArgsDefined(hintDoc);
+    this._hint = hintDoc;
+    return this;
+  }
+
+  remove(): Bulk {
+    this._parentBulk._batchCounts.nRemoveOps++;
+    this._innerFind.remove();
+    return this._parentBulk;
+  }
+
+  removeOne(): Bulk {
+    this._parentBulk._batchCounts.nRemoveOps++;
+    this._innerFind.removeOne();
+    return this._parentBulk;
+  }
+
+  replaceOne(replacement: Document): Bulk {
+    this._parentBulk._batchCounts.nUpdateOps++;
+    assertArgsDefined(replacement);
+    const op = { ...replacement };
+    if (this._hint) {
+      op.hint = this._hint;
+    }
+    this._innerFind.replaceOne(op);
+    return this._parentBulk;
+  }
+
+  updateOne(update: Document): Bulk {
+    this._parentBulk._batchCounts.nUpdateOps++;
+    assertArgsDefined(update);
+    const op = { ...update };
+    if (this._hint) {
+      op.hint = this._hint;
+    }
+    if (this._arrayFilters) {
+      op.arrayFilters = this._arrayFilters;
+    }
+    this._innerFind.updateOne(op);
+    return this._parentBulk;
+  }
+
+  update(update: Document): Bulk {
+    this._parentBulk._batchCounts.nUpdateOps++;
+    assertArgsDefined(update);
+    const op = { ...update };
+    if (this._hint) {
+      op.hint = this._hint;
+    }
+    if (this._arrayFilters) {
+      op.arrayFilters = this._arrayFilters;
+    }
+    this._innerFind.update(op);
+    return this._parentBulk;
+  }
+
+  upsert(): BulkFindOp {
+    assertArgsDefined();
+    this._innerFind.upsert();
+    return this;
+  }
+}
+
+
+@shellApiClassDefault
+@hasAsyncChild
+export default class Bulk extends ShellApiClass {
+  _mongo: Mongo;
+  _collection: any; // to avoid circular ref
+  _batchCounts: any;
+  _executed: boolean;
+  _batches: any;
+  _innerBulk: any;
+
+  constructor(collection, innerBulk) {
+    super();
+    this._collection = collection;
+    this._mongo = collection._mongo;
+    this._innerBulk = innerBulk;
+    this._batches = [];
+    this._batchCounts = {
+      nInsertOps: 0,
+      nUpdateOps: 0,
+      nRemoveOps: 0
+    };
+  }
+
+  /**
+   * Internal method to determine what is printed for this class.
+   */
+  _asPrintable(): string {
```

Comment:
`tojson` looks like is returning an object, is it right that the return type is a string? 

by the way, if you have a type you know is going to be an object and just don't know its  shape you could use `Record`:

ie.
```
tojson(): Record<string, any> { ... }
```

although we could have few utility types for that to use like we've done for `Document`.

If you know which keys, you can restrict that too:

```
type BulkCounters = Record<'nInsertedOps' | 'nInsertedOps' | '...', number>
```

Reply:
Oh cool, yes `Record` will work perfectly for this case. I should be having less `any` in my code, thanks for all the suggestions to improve my typescript!